submissions/kt-std/object-oriented-js/app.js

```diff
@@ -0,0 +1,254 @@
+const allEnemies = [],
+	ENEMIES_AMOUNT = 3,
+	ROW_NUM = 6,
+	COL_NUM = 7,
+	ROW_HEIGHT = 70,
+	Y_STEP = 70,
+	X_STEP = 70,
+	INITIAL_X = X_STEP * Math.floor(COL_NUM / 2),
+	INITIAL_Y = (ROW_NUM - 1) * Y_STEP,
+	CANVAS_START = 0,
+	CANVAS_WIDTH = COL_NUM * X_STEP,
+	CANVAS_HEIGHT = ROW_NUM * X_STEP + 20,
+	LIVES_AMOUNT = 3,
+	LIVE_SPRITE = "images/heart.svg",
+	LIVES_CONTAINER = document.getElementById("hearts"),
+	RANDOM_BASE = 100,
+	SPEED_INC = 0.01,
+	X_GAP = X_STEP / 2,
+	Y_GAP = Y_STEP / 2,
+	messages = {
+		win: {
+			text: `You've just successfully saved Xmas!`,
+			path: `images/bell.svg`,
+			button: `Play again!`,
+			heading: `Congrats!`,
+		},
+		lose: {
+			text: ``,
+			path: `images/bell.svg`,
+			button: `Try again!`,
+			heading: `You lose \u26C4`,
+		},
+	};
+
+class Entity {
+	constructor(x, y, sprite) {
+		this.x = x;
+		this.y = y;
+		this.sprite = sprite;
+	}
+
+	render() {
+		ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+	}
+
+	setPosition(positionX, positionY) {
+		this.x = positionX;
+		this.y = positionY;
+	}
+}
+
+class Enemy extends Entity {
+	constructor(x, y, sprite) {
+		super(x, y, sprite);
+		this.speed = Math.random() * RANDOM_BASE;
+	}
+
+	update(dt) {
+		// Speed up the enemy during the game
+		this.speed += SPEED_INC;
+		// Update Enemy's horizontal position
+		this.x = this.updateHorizontalPosition(this.x + dt * this.speed);
+	}
+	updateHorizontalPosition(position) {
+		/* If the Enemy moves right off the visible canvas area
+		 * place it back to the beginning of the row in other cases
+		 * move right
+		 */
+		return position > CANVAS_WIDTH + X_GAP
+			? CANVAS_START - X_GAP
+			: position;
+	}
+}
+
+class Player extends Entity {
+	constructor(x, y, sprite) {
+		super(x, y, sprite);
+		this.lives = LIVES_AMOUNT;
+	}
+	/* This function resets Player lives to the initial state
+	 * and calls the function to append lives to the LIVES_CONTAINER
+	 */
+	resetLives(livesAmount) {
+		this.lives = livesAmount;
+		appendLivesToContainer(this.lives);
+	}
+
+	/* This function checks if there are left any Player lives.
+	 * If there remain no lives, display a loss message.
+	 * If Player still left lives and it has crossed the border of any Enemy,
+	 * then decrease Player lives amount, delete live item from the LIVES_CONTAINER
+	 * and reset Player position to the initial state.
+	 */
+
+	update() {
+		if (!this.lives) {
+			displayMessage("lose", "flex", messages);
+		} else {
+			allEnemies.forEach((enemy) => {
+				if (
+					Math.abs(this.x - enemy.x) < X_GAP &&
+					Math.abs(this.y - enemy.y) < Y_GAP
+				) {
+					this.lives--;
+					removeItemFromContainer(LIVES_CONTAINER);
+					this.setPosition(INITIAL_X, INITIAL_Y);
+				}
+			});
+		}
+	}
```

Comment:
Separation of Concerns principle should be followed
(e.g. `update` method does only rendering and doesn't contain any unrelated **inline** code; for example collision check is defined as a dedicated method and only called from inside `update`)

Reply:
@OleksiyRudenko, thank you for reviewing and giving vital advices! :) 
Fixed this one via adding one more method to Player class. I do hope the code is better now.